cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java

```diff
@@ -915,6 +918,57 @@ public long getProgramRunCount(ProgramId programId) throws NotFoundException {
     });
   }
 
+  @Override
+  public List<ProgramHistory> getRuns(Collection<ProgramId> programs, ProgramRunStatus status, long startTime,
+                                      long endTime, int limit, Predicate<RunRecordMeta> filter) {
+    return Transactionals.execute(transactional, context -> {
+      List<ProgramHistory> result = new ArrayList<>(programs.size());
+      AppMetadataStore appMetadataStore = getAppMetadataStore(context);
+
+      for (ProgramId program : programs) {
+        ApplicationMeta appMeta = appMetadataStore.getApplication(program.getParent());
+        if (appMeta == null) {
+          result.add(new ProgramHistory(program, Collections.emptyList(),
+                                        new ApplicationNotFoundException(program.getParent())));
+          continue;
+        }
+
+        ApplicationSpecification appSpec = appMeta.getSpec();
+        ProgramType type = program.getType();
```

Comment:
You can use `getExistingAppProgramSpecification()` in this class to get the program spec

Reply:
good suggestion, should have looked around for similar functionality.